components/dm/ansible/import.go

```diff
@@ -0,0 +1,605 @@
+	user    string
+	sources map[string]*SourceConfig // addr(ip:port) -> SourceConfig
+
+	// only use for test.
+	// when setted, we use this executor instead of getting a truly one.
+	testExecutorGetter ExecutorGetter
+}
+
+// NewImporter create an Importer.
+func NewImporter(ansibleDir string, inventoryFileName string) (*Importer, error) {
+	dir, err := filepath.Abs(ansibleDir)
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+
+	return &Importer{
+		dir:               dir,
+		inventoryFileName: inventoryFileName,
+		sources:           make(map[string]*SourceConfig),
+	}, nil
+}
+
+func (im *Importer) getExecutor(host string, port int) (e executor.Executor) {
+	if im.testExecutorGetter != nil {
+		return im.testExecutorGetter.Get(host)
+	}
+
+	keypath := ansible.SSHKeyPath()
+
+	cfg := executor.SSHConfig{
+		Host:    host,
+		Port:    port,
+		User:    im.user,
+		KeyFile: keypath,
+		Timeout: time.Second * time.Duration(im.sshTimeout),
+	}
+
+	e = executor.NewSSHExecutor(cfg, false, im.native)
+
+	return
+}
+
+func (im *Importer) fetchFile(host string, port int, fname string) (data []byte, err error) {
+	e := im.getExecutor(host, port)
+
+	tmp, err := ioutil.TempDir("", "tiup")
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+	defer os.RemoveAll(tmp)
+
+	tmp = filepath.Join(tmp, filepath.Base(fname))
+
+	err = e.Transfer(fname, tmp, true /*download*/)
+	if err != nil {
+		return nil, errors.Annotatef(err, "transfer %s from %s:%d", fname, host, port)
+	}
+
+	data, err = ioutil.ReadFile(tmp)
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+
+	return
+}
+
+func setConfig(config *map[string]interface{}, k string, v interface{}) {
+	if *config == nil {
+		*config = make(map[string]interface{})
+	}
+
+	(*config)[k] = v
+}
+
+// handleWorkerConfig fetch the config file of worker and generate the source
+// which we need for the master.
+func (im *Importer) handleWorkerConfig(srv *spec.WorkerSpec, fname string) error {
+	data, err := im.fetchFile(srv.Host, srv.SSHPort, fname)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	config := new(Config)
+	err = toml.Unmarshal(data, config)
+	if err != nil {
+		return errors.AddStack(err)
+	}
+
+	source := config.ToSource()
+	im.sources[srv.Host+":"+strconv.Itoa(srv.Port)] = source
+
+	return nil
+}
+
+// ScpSourceToMaster scp the source files to master,
+// and set V1SourcePath of the master spec.
+func (im *Importer) ScpSourceToMaster(topo *spec.Topology) (err error) {
+	for i := 0; i < len(topo.Masters); i++ {
+		master := &topo.Masters[i]
+		target := filepath.Join(firstNonEmpty(master.DeployDir, topo.GlobalOptions.DeployDir), "v1source")
+		master.V1SourcePath = target
+
+		e := im.getExecutor(master.Host, master.SSHPort)
+		_, stderr, err := e.Execute("mkdir -p "+target, false)
+		if err != nil {
+			return errors.Annotatef(err, "failed to execute: %s", string(stderr))
+		}
+
+		for addr, source := range im.sources {
+			f, err := ioutil.TempFile("", "tiup-dm-*")
```

Comment:
How about giving it a meaningful name? like `tiup-source-{source-id}.yaml`?

Reply:
as a temporal name, I think there's no need to do this, the target name in the master is '{host}:{port}.yml'(specified at L239, where host and port are the related worker's host and port.

Reply:
OK